lib/response-types/abstract-response-type.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+/**
+ * Module dependencies.
+ */
+
+var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidScopeError = require('../errors/invalid-scope-error');
+var Promise = require('bluebird');
+var is = require('../validator/is');
+var url = require('url');
+
+/**
+ * Constructor.
+ */
+
+function AbstractResponseType(options) {
+  options = options || {};
+
+  if (!options.authorizationCodeLifetime) {
+    throw new InvalidArgumentError('Missing parameter: `authorizationCodeLifetime`');
+  }
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  this.authorizationCodeLifetime = options.authorizationCodeLifetime;
+  this.model = options.model;
+}
+
+/**
+ * Build an error response that redirects the user-agent to the client-provided url.
+ */
+
+AbstractResponseType.prototype.buildErrorRedirectUri = function (redirectUri, { error, ...query } = {}) {
```

Comment:
Would this work as a general purpose redirectUri builder?

```js
AbstractResponseType.prototype.buildRedirectUri = function (redirectUri, { error, ...query } = {}) {
  if (!redirectUri) {
    throw new InvalidArgumentError('Missing parameter: `redirectUri`');
  }

  var uri = url.parse(redirectUri, !error);
  
  uri.query = {
    ...uri.query,
    ...query
  };

  if (error) {
    uri.query.name = uri.query.name || error.name;

    if(error.message) {
      uri.query.error_description = error.message;
    }
  }

  return uri;
};
```

Reply:
In offline discussion with @Americas it was decided to keep the `buildErrorRedirectUri` to match authenticate and token flows. However, it was missing a check for `redirectUri` not being passed to `buildErrorRedirectUri` which has been added now.